docs/design/password-key-derivation.md

```diff
@@ -0,0 +1,77 @@
+# Title: Password Key Derivation
+
+## Abstract
+
+This design doc describes the way cryptographic encryption keys will be derived from user provided passwords in a way that resists offline attacks and low entropy passwords.
+
+## Background
+
+A unique encryption key exists for every path inside of a bucket, and they are designed in a way that allows for deriving a key from a parent and some path component. For example, if you had the key for `sj://bucket/path1/path2` then you could derive the key for `sj://bucket/path1/path2/path3`.
+
+That leaves open the question for how a root key is derived. Some requirements on that root key derivation include
+
+1. It's based on the user's choice (for example, a password).
+2. If the user inputs the same password on different machines, the same root key should be derived.
+
+These requirements allow users to be in full control of their encryption, and don't require users safely transporting high entropy (hard to remember) secrets to bootstrap new uplinks.
+
+This design accomodates more requirements that allow for additional features:
+
+3. A root key can be derived for any path in a bucket.
+4. Buckets root keys should not depend on the name of the bucket.
+5. A table of keys for low entropy passwords should not be possible. In other words, an attacker with knowledge of the algorithm should not be able to use a dictionary of common passwords and pre-compute what keys to check in the event of a data breach.
+6. Users do not have to enter a password for every bucket they create: they can have a default password that is used for every bucket unless otherwise specified.
+
+The third requirement allows having multiple encrypted domains to exist within a single bucket, allowing delegation of encryption. The fourth requirement allows renaming buckets without having to re-encrypt all of the paths inside of a bucket. The fifth requirement enhances security if a satellite breach happens. This is usually accomplished with [salting](https://en.wikipedia.org/wiki/Salt_(cryptography)). The sixth requirement allows providing a finite amount of information to a third party that allows them access to the keys for any bucket created with the default password in the future.
+
+## Design
+
+Satellites will add a `salt` field to bucket metadata that lives with the path encryption information that already exists and that we already request. Satellites will also store a `salt` that lives with the project metadata. The salts will be 32 random, stable bytes.
+
+A default password can be chosen for an api key. The following algorithm is used to derive the default password:
```

Comment:
What is exactly the "default password"? Is it the `ProjectEncryptionKey` from #1959? If yes, can we call it "project encryption key"?

Reply:
Yeah, it would be stored in the `ProjectEncryptionKey`. I agree that the terminology should be consistent.

Reply:
We should probably never use the term "`Key`" to refer to a password/passphrase.

Reply:
We should probably never use the term "Key" to refer to a password/passphrase.

Reply:
Yeah, it's a bit hard for me to come up with what it is: we're taking a password, passing it through a key derivation function, then using it again in the password spot of a key derivation function.

Given that it's the output of a KDF, is it alright to call it a key, even if we use in in the password portion of a KDF later?

Reply:
I should clarify! I meant we should probably never use the term "Key" to refer to something that contains (or might contain) a human-supplied passphrase, which presumably has fairly low entropy. The output of a KDF can certainly be called a Key as far as I'm concerned.

Reply:
OK, perhaps the "default password" terminology in this design is what's misleading then, as it's the output of a KDF.